git-branchless-revset/src/builtins.rs

```diff
@@ -178,8 +180,57 @@ fn fn_heads(ctx: &mut Context, name: &str, args: &[Expr]) -> EvalResult {
 
 #[instrument]
 fn fn_branches(ctx: &mut Context, name: &str, args: &[Expr]) -> EvalResult {
-    eval0(ctx, name, args)?;
-    Ok(ctx.dag.branch_commits.clone())
+    let pattern = match eval0_or_1_pattern(ctx, name, args)? {
+        Some(pattern) => pattern,
+        None => return Ok(ctx.dag.branch_commits.clone()),
+    };
+
+    let all_branch_commits: HashSet<_> = ctx
+        .dag
+        .commit_set_to_vec(&ctx.dag.branch_commits)
+        .map_err(EvalError::OtherError)?
+        .into_iter()
+        .collect();
+    let branch_commits = make_pattern_matcher(
+        ctx,
+        name,
+        args,
+        Box::new(move |repo: &Repo, commit: &Commit| {
+            if !all_branch_commits.contains(&commit.get_oid()) {
+                return Ok(false);
+            }
```

Comment:
This is my workaround for not being able to restrict the commits that the `Matcher` will run on from the default of all visible commits to what we really need, which is just the branch commits. Refactoring `make_pattern_matcher()` etc felt like more that was needed for this limited use case.

Happy to reconsider if you think that would be useful, though.

Reply:
> This was my misunderstanding of how revset Matchers work. I hadn't connected that they work on all visible commits, and there is currently no way to restrict that. (This is the first revset feature – that I can tell – that would make any sense to operate on anything other than all visible commits, because we know that branches are (probably) a subset of visible commits.)

Take a look at `CommitSet::from_evaluate_contains`/`CommitSet::from_async_evaluate_contains`. For the branches case, we want to provide separate paths for `evaluate` vs `contains`. The `Matcher` stuff is just for the case where we want to provide only `contains` and define `evaluate` to mean "run `contains` on all commits in the repo".

In principle, if we provide `Hints::FILTER`, then the runtime library should know how to optimize e.g. `merges() & expr()` so that we can try to evaluate `expr()` first if it seems to be more efficient to do so, but I don't know if/how this actually happens in practice. If the library doesn't do that, then perhaps we should do the check ourselves in our intersection implementation and swap the operands. Possibly something like this —

```rust
#[instrument]
fn fn_intersection(ctx: &mut Context, name: &str, args: &[Expr]) -> EvalResult {
    let (lhs, rhs) = eval2(ctx, name, args)?;
    let (lhs, rhs) = if rhs.hints().contains(Flags::FILTER) && !lhs.hints().contains(Flags::FILTER)
    {
        (rhs, lhs)
    } else {
        (lhs, rhs)
    };
    Ok(lhs.intersection(&rhs))
}
```

As written, I think that this pessimizes the zero-argument case of `branches()`, because now we have to traverse all visible commits to see if each one points to a branch. We should be able to get both contains and evaluate paths to be efficient.

Reply:
> pessimizes the zero-argument case of branches(),

I'm probably misunderstanding you, but I think that the zero-argument case is nearly identical to the same case on `master`: it just immediately returns `ctx.dag.brance_commits`. When an arg is passed, though, then – yes – it does have to traverse all visible commits to find those at matching branches.

> Take a look at CommitSet::from_evaluate_contains/CommitSet::from_async_evaluate_contains

OK, cool, but I'm still having a hard time pulling apart how to best make use of them. For now, I've made the following updates:
- `merges()` now uses the pattern matcher code (w/o a pattern matcher, though) to leverage the `evalulate`/`contains` paths w/ a very simple filtering fn. This revset is now basically just 1 line of code.
- `branches()` I removed the hacky `HashSet` thing I was doing and instead altered the API of `make_pattern_matcher_set()` so that it can (optionally) take a set of commits to traverse instead of just "all visible commits"
  - this made sense because we already know the commits we need to examine, via `dag.branch_commits`
  - another option I considered was to use `repo.get_all_local_branches()`, and then using the `pattern` to filter by name, finally resolving the branches to a set via `branch.get_oid()` ... this seemed like it may be less work on our end (more direct), but I didn't explore it because the current implementation is at least working 😄 

But I may also just be misunderstanding what you're suggesting! Sorry, if so.

> For the branches case, we want to provide separate paths for evaluate vs contains.

I wonder if my change (to pass the branch commits into `make_pattern_matcher_set()`) makes this less of an issue, because now the `evaluate()` path will not operate on all visible commits? If I'm misunderstanding what you're asking for, please let me know and I'll dig in again.

> If the library doesn't do that, then perhaps we should do the check ourselves in our intersection implementation and swap the operands

I checked briefly and it was unclear if the upstream is doing this level of optimization (though I'd be a bit surprised if that's the case). They seem mostly concerned if either side is `FULL` or `EMPTY`, but I did not dig too deeply. (I found [this](https://github.com/facebook/sapling/blob/fa6824a3add28a36d7860dc91d8534f08b3941d8/eden/scm/lib/dag/src/nameset/mod.rs#L228) and [this](https://github.com/facebook/sapling/blob/main/eden/scm/lib/dag/src/nameset/intersection.rs#L92); the latter suggests that they prefer to keep `FILTER` sets on the rhs?)

Regardless, I can look at this, but I think I would prefer to do so in a separate PR, unless you think the perf of this PR is really not good. I may need a suggestion on how to benchmark that, though, as most of the repos I have access to are pretty fast already! 😆

Reply:
> another option I considered was to use repo.get_all_local_branches(), and then using the pattern to filter by name, finally resolving the branches to a set via branch.get_oid()

FWIW I did end up playing with this, and this seems to work, and with a lot less code!

```rust
fn fn_branches(ctx: &mut Context, name: &str, args: &[Expr]) -> EvalResult {
    let pattern = match eval0_or_1_pattern(ctx, name, args)? {
        Some(pattern) => pattern,
        None => return Ok(ctx.dag.branch_commits.clone()),
    };

    let branch_commits = ctx
        .repo
        .get_all_local_branches()?
        .iter()
        .filter(|branch| pattern.matches_text(branch.get_name().expect("TODO")))
        .filter_map(|branch| branch.get_oid().expect("TODO never happens"))
        .collect();

    Ok(branch_commits)
}
```

Reply:
I think specifying the universe of commits in the pattern matcher should work performantly.